src/java.base/share/native/libjava/io_util.c

```diff
@@ -53,6 +53,63 @@ readSingle(JNIEnv *env, jobject this, jfieldID fid) {
     return ret & 0xFF;
 }
 
+#define SHORT_SIZE 2
+#define INT_SIZE   4
+#define LONG_SIZE  8
+
+jlong
+readN(JNIEnv *env, jobject this, jfieldID fid, jint len) {
+    if (len == 1) {
+        return readSingle(env, this, fid);
+    }
+    char *buf = NULL;
+    if (len == SHORT_SIZE) {
+      char stackBuf[SHORT_SIZE];
+      buf = stackBuf;
+    }
+    if (len == INT_SIZE) {
+      char stackBuf[INT_SIZE];
+      buf = stackBuf;
+    }
+    if (len == LONG_SIZE) {
+      char stackBuf[LONG_SIZE];
+      buf = stackBuf;
+    }
+
+    FD fd = getFD(env, this, fid);
+    if (fd == -1) {
+        JNU_ThrowIOException(env, "Stream Closed");
+        return -1;
+    }
+    jint nread = IO_Read(fd, buf, len);
+    if (nread == -1) {
+        JNU_ThrowByName(env, "java/io/EOFException", NULL);
+        return -1;
+    }
+
+    if (len == SHORT_SIZE) {
+        return ((*buf) << 8) + (*(buf + 1) & 0xFF);
+    }
+    if (len == INT_SIZE) {
+        return ((*buf) << 24) +
+               ((*(buf + 1) & 0xFF) << 16) +
+               ((*(buf + 2) & 0xFF) << 8) +
+               ((*(buf + 3) & 0xFF));
+    }
+    if (len == LONG_SIZE) {
+        return ((((jlong) *(buf)) << 56) +
+                (((jlong) (*(buf + 1)) & 0xFF) << 48) +
+                (((jlong) (*(buf + 2)) & 0xFF) << 40) +
+                (((jlong) (*(buf + 3)) & 0xFF) << 32) +
+                (((jlong) (*(buf + 4)) & 0xFF) << 24) +
+                ((*(buf + 5) & 0xFF) << 16) +
+                ((*(buf + 6) & 0xFF) << 8) +
+                ((*(buf + 7) & 0xFF)));
+    }
```

Comment:
Doing this in a JNI method is problematic for several reasons and I think we should try to keep the JNI methods as simple/short as possible, ideally map to a single syscall and not do too much if possible. In the medium/long term then many of the native methods in this area will be replaced anyway, maybe by re-implementing RAF/etc. on NIO, maybe by replacing the JNI code with the java.lang.foreign APIs. In addition, I expect there will be another tranche of changes that will significantly change this code for virtual threads to allow the underlying thread be released when doing for file I/O.

Reply:
So I should roll the changes to VM back and modify only Java code, right?